<a href="https://colab.research.google.com/github/Mogi83/CS-repo/blob/main/CS%20133/Homework/homework-01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 1 - Introduction to Python
CS 133  
Spring 2025  
v1.0

---

### 1. Data types

Python is a _strongly_ typed language because data types do not change in Python, i.e. a *string* is always a string and an *int* is always an integer. To change a data type requires an explicit operation (called a _cast_).

However, unlike many programming languages Python is _dynamically_ typed, **not** _statically_ typed. This means the data type (*string*, *int*, etc.) is enforced on _values_ not _variables_. Variables in Python are simply loose references to data values (also known as _objects_) and can be changed to refer to different objects at any time (even if those objects have different types). That can be confusing if you're used to conventional programming languages.

Python has many features that make it expressive and concise. One of these is doing multiple variable assignments in a single statement. For example, to initialize the variables `x`, `y`, and `z` all to 0 you could write:

```x = y = z = 0```

Technically, what this is doing is first assigning 0 to the variable `z`, then assigning whatever is in the variable `z` to `y`, and so on. This is because the order of precedence for the assignment operator goes 'inside out'.

1.1 (1 pt) Create two variables named `a` and `b`, and use a **single statement** to assign `b` to the integer `3` and `a` to the variable `b`

In [3]:
a = b = 3
print (a, b)

3 3


1.1.2 (1 pt) Change the value of the `b` variable to the float `5.0` then state the value of the `a` variable. (Note: to **state** the value of a variable, just put it on a line by itself - Python will output its value to the console)

In [ ]:
b = 5.0
print ("a: " + str(a), "b: " + str(b))

a: 3 b: 5.0


1.1.3 (1 pt) You assigned `b` to the `a` variable in 1.1 above. Then you changed the value of the `b` variable. Why didn't the value of `a` also change?

In code block [1] both named variables 'a' and 'b' are pointing to an integer object with the value of 3. However in [2] 'b' points to a new float object, because 'a' is not being reassigned to equal 'b' (like in block [1]) it continues to point to its originally assigned value of 3.

1.1.4 (1 pt) Now set the `a` and `b` variables to the same list `['c', 'h', 'a', 'n', 'c', 'e' ]`

In [4]:
a = b = ['c', 'h', 'a', 'n', 'c', 'e' ]

1.1.5 (1 pt) Using the `b` variable, change the value of the fifth letter in the list to `g`

In [5]:
b[5] = 'g'
print (b)

['c', 'h', 'a', 'n', 'c', 'g']


1.1.5 (1 pt) State the value of `a` and `b`. (Note: only the output of the last line of a cell will be shown. To state multiple variables put them on the same line separated by commas)

In [ ]:
print (a)
print(b)

['c', 'h', 'a', 'n', 'c', 'g']
['c', 'h', 'a', 'n', 'c', 'g']


1.1.6 (1 pt) This is a similar pattern to what you did in question 1.1. Why did the value of `a` change this time?

I assigned a equal to b in the start as they have the same assignment of the array.

### 2. Tuples

Tuples are _immutable_ lists that are very useful in Python. They can be written with or without parentheses (in some cases) but it is usually good practice to include the parentheses.

You can use tuples in assignment statements on both sides of the `=` operator and Python will order the values appropriately.

2.1 (2 pts) Assign the variables `x`, `y`, and `z` to the values `True`, `0`, and `'blue'` using tuples in a single statement. Show the values of `x`, `y`, and `z` without using the `print()` function by stating them in tuple form.

In [ ]:
x, y, z = (True, 0, 'blue')
(x, y, z)

(True, 0, 'blue')

You access elements in a tuple just like a list, i.e. using the `[n]` notation, where `n` is the `n+1st` element of the tuple.

2.2 (1 pt) Assign a variable named `seq` a tuple of values `100`, `101`, `102`, `103`, and `103`.

In [ ]:
seq = (100,101,102,103,103)

2.2.1 (1 pt) State the `103` element of `seq` by referencing it using the `[n]` notation

In [ ]:
seq[4]

103

2.3 (5 pts) Using a tuple to represent the position of player's token in a 3-d grid, create a function that will move the player's token in one of six directions (along the three axes). Name the function `move(token, dir)` where `token` is a tuple of `(x,y,z)` corresponding to the token's position and `dir` is one of (x-axis: `right`, `left`), (y-axis: `up`, `down`), (z-axis: `forward`, `backward`). The function should return a new tuple with the coordinates of the new position if `dir` is a recognized direction, otherwise it should return the original position.

Example use: `move( (0,0,0), 'forward')` should return the value `(0,0,1)`

In [ ]:
dir_map = {
    #x-axis
    "right": (1, 0, 0),
    "left": (-1, 0, 0),

    #y-axis
    "up": (0, 1, 0),
    "down": (0, -1, 0),

    #z-axis
    "forward": (0, 0, 1),
    "backward": (0, 0, -1)
}


def move(token, dir):
    new_pos_list = list(token)
    if dir in dir_map:
        adjustment = dir_map[dir]

        for i in range(len(new_pos_list)):
            new_pos_list[i] += adjustment[i]

        return tuple(new_pos_list)
    else:
        return token

2.4.1 (3 pts) Use a series of calls to your function to move a token from the position (2, 3, -1) back to the origin (0, 0, 0). Be sure to print the final value by stating it.

In [ ]:
givenPos = (2, 3, -1)

givenPos = move(givenPos, "left")
givenPos = move(givenPos, "left")
#expected value (0, 3, -1)
#print(givenPos)


givenPos = move(givenPos, "down")
givenPos = move(givenPos, "down")
givenPos = move(givenPos, "down")
#expected value (0, 0, -1)
#print(givenPos)


givenPos = move(givenPos, "forward")
#finished, should be (0, 0, 0)

givenPos

(0, 0, 0)

### 3. Lists, Dictionaries, and Sets

Lists are similar to tuples but are _mutable_, meaning they can be changed (you can delete or add elements to a mutable list). Lists are ordered so you must access elements using the indexed notation `[n]`.  

Dictionaries are not ordered. Instead you access an element using a `key`. The key can be any hashable data type (e.g. strings, integers, etc.) but all keys used in a dictionary must be unique. The `key` is like a label or name of a value in the dictionary. You can retrieve a value from the dictionary using a syntax similar to indexing in a list, i.e. `[key]` where `key` is either a constant value or a variable containing a value.  

Sets are like dictionaries without keys. The values stored in a set are unordered like dictionaries, but you cannot access them with a `key` or index. Only one copy of a value will be kept for a set which means every value in a set will be unique.

3.1 (2 pts) Create a variable and assign it a list containing the names of the seven colors of the rainbow (red, orange, yellow, blue, green, indigo, violet). State the value of the third element in the list referenced by the variable.

In [ ]:
rainbow = ['red', 'orange', 'yellow', 'blue', 'green', 'indigo', 'violet']
rainbow[2]
rainbow

['red', 'orange', 'yellow', 'blue', 'green', 'indigo', 'violet']

3.1.1 (2 pts) Use the list method `insert(position, value)` to add the color `black` to the front of the list. State the new value of the list variable.

In [ ]:
rainbow.insert(0, 'black')
rainbow

['black', 'red', 'orange', 'yellow', 'blue', 'green', 'indigo', 'violet']

3.1.2 (2 pts) Use the list method `append(value)` to add the color `white` to the back of the list. State the new value of the list variable.

In [ ]:
rainbow.append('white')
rainbow

['black',
 'red',
 'orange',
 'yellow',
 'blue',
 'green',
 'indigo',
 'violet',
 'white']

3.2 (3 pts) Create a variable containing an empty dictionary. Add three items to the dictionary with the following keys: `name`, `age`, and `occupation`. The corresponding values should be `Bob`, `27`, and `rocket scientist`. Use the variable to state the value of the key `age`.

In [ ]:
newDict = {

}
newDict.update({"name": "Bob", "age": "27", "occupation": "rocket scientest"})

newDict["age"]

'27'

3.2.1 (1 pt) What happens if you try to access a key that does not exist in the dictionary?

We would get a KeyError.

3.2.2 (2 pts) Create a variable containing a `set` with the first six digits of pi (3, 1, 4, 1, 5, 9). State the value of the variable.

In [ ]:
pi = {3, 1, 4, 1, 5, 9}
pi

{1, 3, 4, 5, 9}

3.2.3 (1 pt) Why did the value change compared to it's initial value?

Sets don't repeat values, so even though the object was assigned with two values of 1 we only state it once.

3.3 (3 pts) You are a game designer and need a way to generate characters for your game. Create the following three lists:  

- character types: include three or more types, e.g. warrior, wizard, thief
- weapons: include three or more weapons, e.g. sword, bow, wand
- abilities: include three or more special abilities, e.g. fireball, stealth

In [ ]:
char_type = ['Fighter', 'Ranger', 'Bard', 'Wizard', 'Sorcerer', 'Warlock', 'Rouge']
weap_type = ['Longsword', 'Shortsword', 'Dagger', 'Mace', 'Lute', 'Spellbook', 'Longbow', 'Shortbow']
abil_type = ['Slash', 'Parry','Second Wind', 'Precise Shot', 'Hidden Strike', 'Shield Bash', 'Fireball', 'Invisibility']

#Test values for duplicate classes
#char_type = ['Fighter']
#weap_type = ['Longsword']
#abil_type = ['Slash']

3.3.1 (10 pts) Use a loop to create five unique characters by randomly choosing one element from each of the three lists. For each new character create a dictionary that stores the generated values. Add each new character to a list but first check if a duplicate exists by creating a string from the three values and checking if the string exists in a set. If not, store the string in the set and add the new character to the character list.

You can use the following pseudo code if you wish:
- import the `random` module
- initialize a variable with an empty list to hold the new characters
- initialize a variable to an empty set to use for checking duplicates (note you will need to use the `set()` constructor)
- loop 5 times, e.g. `for i in range(5)`
    - initialize a temporary empty dictionary to hold a new character
    - set the new character's type to a random value of the character type list (hint: use `random.choice(ctype)` where `ctype` is the name of your list)
    - repeat the step above for weapon and ability
    - convert the new character into a string using `f"{character}"` where `character` is the variable name you created to temporarily hold the new character
    - check if the string is contained in the set you created for checking duplicates (hint: use the syntax `x in y` where `x` is the element and `y` is the set)
    - if the string is not in the set
        - add the string to the set
        - append the new character to the list of new chracters

Make sure to print your new character list by stating it at the end of the cell

In [ ]:
import random

char_list= []
duplicate_check = set({})

def characterGenerator (num = 5):
    for i in range(num):
        temp_char = {

        }

        temp_char.update({"type": random.choice(char_type), "weapon": random.choice(weap_type), "ability": random.choice(abil_type)})
        if f"{temp_char}" in duplicate_check:
            print("Character already exists!")
        else:
            duplicate_check.add(f"{temp_char}")
            char_list.append(temp_char)


    print(char_list)


characterGenerator()

[{'type': 'Ranger', 'weapon': 'Spellbook', 'ability': 'Invisibility'}, {'type': 'Warlock', 'weapon': 'Longbow', 'ability': 'Parry'}, {'type': 'Rouge', 'weapon': 'Shortsword', 'ability': 'Slash'}, {'type': 'Sorcerer', 'weapon': 'Longsword', 'ability': 'Parry'}, {'type': 'Sorcerer', 'weapon': 'Shortsword', 'ability': 'Hidden Strike'}]


### 4. Functions

4.1 (10 pts) Create a function named `recipe_scaler` that takes the following parameters:  

`ingredients` : a dictionary that maps ingredient names to amounts (note these amounts will be for 1 serving and will be specified in ounces with the exception of eggs)  
`servings` : the number of servings to scale (default should be 1 which means no scaling)  
`units` : the type of units (default should be `metric` - all input values to the function will be in imperial ounces)  

The function should scale the recipe by the number of `servings` passed in, convert between metric and imperial units if needed, and return a new dictionary with the adjusted ingredients.

When converting between metric and imperial use 1 oz = 28.35 g. However, make sure to handle the case for eggs (no conversion).

Example output:  

```recipe_scaler( { 'flour': 2, 'sugar': 1, 'eggs': 3 }, 2, 'imperial') => { 'flour': 113.4, 'sugar': 56.7, 'eggs': 6 }```

In [ ]:
def recipe_scaler(ingredients: dict, servings = 1, units = "metric"):
    scaled_ingredients = {

    }

    ingredient_names = list(ingredients.keys())
    ingredient_values = list(ingredients.values())

    for i in range(len(ingredient_names)):
        if units == 'imperial' and ingredient_names[i] != 'eggs':
            scaled_ingredients.update({ingredient_names[i]: (ingredient_values[i] * 28.35 * servings)})
        elif ingredient_names[i] == 'eggs':
            scaled_ingredients.update({'eggs': (ingredient_values[i] * servings)})
        else:
            scaled_ingredients.update({ingredient_names[i]: (ingredient_values[i] * servings)})
    return scaled_ingredients

print(recipe_scaler( { 'flour': 2, 'sugar': 1, 'eggs': 3 }, 2, 'imperial'), "\n")
print(recipe_scaler( { 'flour': 1, 'sugar': 1, 'eggs': 2 }, 2, 'metric'))

{'flour': 113.4, 'sugar': 56.7, 'eggs': 6} 

{'flour': 2, 'sugar': 2, 'eggs': 4}


### 5. Loops and list comprehensions

Besides standard `for` and `while` loops, Python allows you to loop over a list in an expression that can be used to generate new lists.

5.1 (3 pts) The `range()` function returns a list of the numbers in a sequence. Use a standard `for` loop using the `range()` function to create a list named `dbt` of the numbers from 0 to 21 where the value is 0 if the number is not divisible by 3. State the value of `dbt` at the end.

In [ ]:
dbt = []
for i in range(22):
    if i % 3:
        dbt.append(0)
    else:
        dbt.append(i)

dbt

[0, 0, 0, 3, 0, 0, 6, 0, 0, 9, 0, 0, 12, 0, 0, 15, 0, 0, 18, 0, 0, 21]

To use a `for` loop expression requires the form  
```
<value> for <var> in <list>
```

In this expression `<value>` is some expression that can evaluate to a constant value (usually in terms of the loop variable). The `<var>` is the loop variable (you can choose whatever you like), and the `<list>` is the list you want to loop over.

Try to guess the output of the cells below and then run them to see if you're correct.

In [ ]:
[ 9 for i in range(9) ]

[9, 9, 9, 9, 9, 9, 9, 9, 9]

In [ ]:
[ i for i in range(9) ]

[0, 1, 2, 3, 4, 5, 6, 7, 8]

In [ ]:
[ i * 2 for i in range(9) ]

[0, 2, 4, 6, 8, 10, 12, 14, 16]

In [ ]:
[ f"{x}ish" for x in [ 'red', 'blue', 'green', 'yellow' ] ]

['redish', 'blueish', 'greenish', 'yellowish']

5.1.2 (2 pts) Create a list using an expresion that generates the numbers from 0 to 21 using list comprehension (don't assign the result to a variable so it will be output to the console)

In [ ]:
[i for i in range(22)]

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]

5.1.3 (3 pts) Python also allows the `if/else` construction inan expression making it useful when combined with a list comprehension. The syntax is:  

```<value> for <var> in <list> if <condition> else <othervalue>```

`<condition>` should evaluate to a boolean value and `<othervalue>` will be used if the `<condition>` is `False`

Start with your expression above and modify it to include an `if/else` expression to set the value to 0 if it is not divisible by 3.

In [ ]:
dbt = [i if (i % 3 == 0) else 0 for i in range(22)]
dbt

[0, 0, 0, 3, 0, 0, 6, 0, 0, 9, 0, 0, 12, 0, 0, 15, 0, 0, 18, 0, 0, 21]

### 6. Numpy

The Numpy package provides a powerful set of tools for working with mult-dimensional arrays that forms the basis for more advanced data science packages.  

A Numpy array is very similar to a list but has many more features. Numpy arrays can be multi-dimensional and this is why they are named `ndarray` (n-dimensional array).  

It is customary to import the numpy package as np as shown below. Run the cell so the package is available for the following problems.

In [ ]:
import numpy as np

6.1 (2 pts) Use the expression you wrote in 5.1.3 to initialize a numpy ndarray using the `array()` function. Assign it to a variable named `threes`.

In [ ]:
threes = np.array([i if (i % 3 == 0) else 0 for i in range(22)])

6.1.2 (1 pt) Display the shape of the `threes` variable by stating the `shape` member of the `threes` array

In [ ]:
threes.shape

(22,)

6.1.3 (1 pt) What does the number(s) in the output above mean?

6.1.4 (2 pts) Reshape the array so that it is 7 by 3 (7 rows of 3 values), using the `reshape()` method. Make sure to reassign `threes` to the result and state the value of `threes`

In [ ]:
#Given that my code above makes a 1 dimension array 22 long when I belive we should have one thats 21 long so I am going to delete my first zero so I can reshape as intended.

threes = np.delete(threes, 0)
print(threes)
print(" ")
threes = threes.reshape((7, 3))
print(threes)

[ 0  0  3  0  0  6  0  0  9  0  0 12  0  0 15  0  0 18  0  0 21]
 
[[ 0  0  3]
 [ 0  0  6]
 [ 0  0  9]
 [ 0  0 12]
 [ 0  0 15]
 [ 0  0 18]
 [ 0  0 21]]


6.2 (2 pts) The `delete(arr, slice, axis)` method can be used to remove a portion of an array. The parameter `slice` can be the index of a single value or a 'slice' of the array. The named parameter `axis` is an integer index of the axis (row=0, col=1, etc.) and defaults to `None` which flattens the array.  

Delete the second column of the `threes` array using the `delete()` method. Be sure to reassign `threes` to the result and state the new value.

In [ ]:
threes = np.delete(threes, 1, axis = 1)
threes

array([[ 0,  3],
       [ 0,  6],
       [ 0,  9],
       [ 0, 12],
       [ 0, 15],
       [ 0, 18],
       [ 0, 21]])

6.3 (5 pts) One of the powerful features of numpy arrays is the ability to perform operations on each element in the array with a simple expression. When a numpy `ndarray` data type is used in an expression, such as a mathematical operation, it will automatically loop over each element in the array and perform the operation on it.  

Numpy arrays can use familiar list indexing to access elements, with commas to denote each access. For example, `threes[2, 1]` will access the element in the 3rd row and 2nd column.  

When a 'slice' is used in the index it stands for 'all values included in the slice'. For example, `threes[3:5, 1]` will access the elements in the 2nd column for all rows between 3 and 5.  

Using the `:` notation alone is shorthand for all rows: `threes[3,:]` gives all elements on row 3 (as a new array).

Set the second column of the `threes` array to the square of the first (use `x ** 2` to square a number).

First try writing an expression that gives only the first element of each row.

In [ ]:
#1st element of each row
print ("First row elements: \n" + str(threes[:, 0]))
threes
print() #for clarity
#**2 column 2 results in not a real change im going to replace the zeros with 1-7 values so we can see what is happening.
threes[:, 0] = range(1, 8)
threes[:, 1] = threes[:, 1] ** threes[:,0]
threes

First row elements: 
[0 0 0 0 0 0 0]



array([[         1,          3],
       [         2,         36],
       [         3,        729],
       [         4,      20736],
       [         5,     759375],
       [         6,   34012224],
       [         7, 1801088541]])

6.3.1 (5 pts) Using the expression above, assign the square of the element in the first column to the second column for each row. State the result

In [ ]:
# Square index 0

threes[:, 1] = threes[:, 0] ** 2
threes

array([[ 1,  1],
       [ 2,  4],
       [ 3,  9],
       [ 4, 16],
       [ 5, 25],
       [ 6, 36],
       [ 7, 49]])

### 7 Pandas

The Pandas library is the next important data science library we will use. It extends the basic Numpy `ndarray` to something called a `DataFrame`, which is a matrix comprised of rows and columns. Pandas is targeted to data science and machine learning so it's features and structure are tailored for that purpose.  

In data science datasets have a series of 'instances' or 'samples'. Each instance is comprised of the values for a set of 'features'. In Pandas each instance is a row in the DataFrame and the features are the columns, much like you would use in a spreadsheet.

Like the Numpy module, Pandas has a powerful set of features that can perform most of the common operations necessary for data science. The key to these operations is the labels Pandas uses. Labels for the rows are called indices and labels for the columns are called column labels.

Run the cell below to use the pandas module in the following exercises

In [ ]:
import pandas as pd

7.1 (4 pts) Create a Pandas `DataFrame` object using the `DataFrame()` constructor and assign it to a variable called `df`. The columns should be `[ color, shape, quantity ]` and there should be at least 3 rows. You can make up the values. Be sure to state the value of `df`.

In [ ]:
df = pd.DataFrame(
    data=[['red', 'triangle', '12'],
          ['blue', 'square', '5'],
          ['yellow', 'circle', '115']],
    columns=['color', 'shape', 'quantity'])

df

color     shape quantity
0     red  triangle       12
1    blue    square        5
2  yellow    circle      115

7.2 (3 pts) Use the `sort_values()` function and choose a column that will change the order of the rows in your dataframe. Be sure to use the `inplace=True` parameter so your dataframe is modified in place (or reassign it to the result of `sort_values()`), then state the vlaue of `df`

In [ ]:
df.sort_values('shape', inplace=True)
df


color     shape quantity
2  yellow    circle      115
1    blue    square        5
0     red  triangle       12

7.2.1 (2 pts) Why is the row index now out of order?

By default pandas.Dataframe.sort_values will sort numerically or alphabeticlly depending on the data type and the column selected, this default is in ascending order. For example if row 1 (index 1) had the value circle and row 0 (index 2) had the value square for their shape values .sort_values would output index 1, 2, 0.

Like the numpy operators, you can use a dataframe in an expression and Pandas will perform the operation on each element.

7.3 (3 pts) Use an expression to multiply each element in the quantity column by 3 and assign it back to the quantity column. State the value of `df`.

In [ ]:
#OLD CODE it works but uhh not a single 'expression' nor is it very elegant. I had to review pandas inbetween this function and the solution below.
#for i in range(len(df.values)):
#    print(df.at[i, 'quantity'])
#    df.replace(to_replace = df.at[i, 'quantity'], value = (int(df.at[i, 'quantity']) * 3), inplace = True )
#    print(df.at[i, 'quantity'])

#the real solution because of course its that easy this is what I get for skipping numpy section because I couldn't get .reshape() to work on my 22 element array.

df['quantity'] = [int(i) * 3 for i in df['quantity']] #goooo list comprehension :)

df

color     shape  quantity
2  yellow    circle       345
1    blue    square        15
0     red  triangle        36

7.4 (3 pts) One of the most common ways to work with a dataframe is to select a subset of rows or column based on some condition. To understand how this works, first use a simple expression to select two of the rows in your dataframe (use a condition on the quantity colum that holds true for only two rows, such as `df.quantity < 50` ). Assign the result to a variable called `selection` and state the value.

In [ ]:
selection = df['quantity'] < 50
selection

2    False
1     True
0     True
Name: quantity, dtype: bool

7.4.2 (1 pt) What is returned from the expression?

7.4.3 (3 pts) Use the variable `selection` to index the `df` dataframe (do not assign the result to a variable)

In [ ]:
df[selection]

color     shape  quantity
1  blue    square        15
0   red  triangle        36

7.4.4 (1 pt) Which rows in `df` were selected (relative to `selection`)?

Only in rows where that index's quantity value was < 50. So index 2 (see above) with a value of 345 was false.

7.4.5 (1 pt) How do you think the row index was used in the above operation?

The row index kept our data together, even though we were looking specifically at the quantity and using expressions on that we never lost the data related to color or shape as it was tied to our row index. Essentially all our series did (selection) was to say add this index or that index to the dataframe.

---

### Submission Instructions

Be sure to ***SAVE YOUR WORK***!  

Then submit this notebook (.ipynb) file to Canvas.